File: lib/resource/edit.php
<?php
/*
Copyright (©) 2003-2015 Teus Benschop.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/


require_once ("../bootstrap/bootstrap");
page_access_level (Filter_Roles::manager ());
$header = new Assets_Header (gettext("Resource"));
$header->run ();
$view = new Assets_View (__FILE__);


$database_resources = Database_Resources::getInstance ();


@$name = request->query['name'];
$view.set_variable ("name = $name;


if (isset(request->post['code'])) {
  $code = request->post['code'];
  include ("config/open");
  if ($open_installation) {
    echo "<h2>Cannot save data in this open installation</h2>";
  } else {
    $database_resources->save ($name, $code);
  }
  $book = request->post ['book'];
  $chapter = request->post ['chapter'];
  $verse = request->post ['verse'];
  $view.set_variable ("posted = true;
}


$file = $database_resources->getFile ($name);
$code = filter_url_file_get_contents ($file);
$view.set_variable ("code = $code;


$database_books = Database_Books::getInstance ();
$books = Database_Books::getIDs ();
$booknames = array ();
for ($books as $id) {
  $booknames [] = Database_Books::getEnglishFromId ($id);
}
$view.set_variable ("books = $books;
$view.set_variable ("booknames = $booknames;


if (!isset ($book)) $book = 1;
if (!isset ($chapter)) $chapter = 1;
if (!isset ($verse)) $verse = 1;
$view.set_variable ("book = $book;
$view.set_variable ("chapter = $chapter;
$view.set_variable ("verse = $verse;


$output = Resource_Logic::getExternal ($name, book, chapter, verse, false);
$view.set_variable ("output = $output;


$file = $database_resources->getFile ($name);
$file = basename ($file);
$view.set_variable ("file = $file;
$metafile = $database_resources->getMetaFile ($name);
$metafile = basename ($metafile);
$view.set_variable ("metafile = $metafile;


$view->render ("edit");


Assets_Page::footer ();


?>

File: lib/notes/edit.cpp
/*
 Copyright (©) 2003-2015 Teus Benschop.
 
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */


#include <notes/edit.h>
#include <assets/view.h>
#include <assets/page.h>
#include <assets/header.h>
#include <filter/roles.h>
#include <filter/string.h>
#include <filter/url.h>
#include <webserver/request.h>
#include <locale/translate.h>
#include <database/notes.h>
#include <notes/logic.h>
#include <access/bible.h>
#include <ipc/focus.h>
#include <navigation/passage.h>
#include <notes/note.h>


string notes_edit_url ()
{
  return "notes/edit";
}


bool notes_edit_acl (void * webserver_request)
{
  return Filter_Roles::access_control (webserver_request, Filter_Roles::manager ());
}


string notes_edit (void * webserver_request)
{
  Webserver_Request * request = (Webserver_Request *) webserver_request;
  Database_Notes database_notes = Database_Notes (webserver_request);
  Notes_Logic notes_logic = Notes_Logic (webserver_request);
  
  
  string page;
  Assets_Header header = Assets_Header (translate("Edit Note Source"), request);
  page += header.run ();
  Assets_View view = Assets_View ();
  
  
  int identifier;
  if (request->query.count ("identifier")) identifier = convert_to_int (request->query ["identifier"]);
  else identifier = convert_to_int (request->post ["identifier"]);
  if (identifier) view.set_variable ("identifier", convert_to_string (identifier));
  
  
  
  if (request->post.count ("data")) {
    // Save note.
    string noteData = request->post["data"];
    if (database_notes.identifierExists (identifier)) {
      database_notes.setContents (identifier, noteData);
      string url = filter_url_build_http_query (notes_note_url (), "id", convert_to_string (identifier));
      // View the updated note.
      redirect_browser (request, url);
      return "";
    }
  }
  
  
  if (identifier) {
    if (database_notes.identifierExists (identifier)) {
      string noteData = database_notes.getContents (identifier);
      view.set_variable ("data", noteData);
    }
  }
  
  
  page += view.render ("notes", "edit");
  page += Assets_Page::footer ();
  return page;
}
